libpromises/policy.c

```diff
@@ -1425,6 +1430,17 @@ Constraint *PromiseAppendConstraint(Promise *pp, const char *lval, Rval rval, bo
         Constraint *old_cp = SeqAt(pp->conlist, i);
         if (strcmp(old_cp->lval, lval) == 0)
         {
+            if (strcmp(old_cp->lval, "ifvarclass") == 0 ||
+                strcmp(old_cp->lval, "if") == 0)
+            {
+                Buffer *grow = BufferNew();
+                BufferAppendF(grow, "(%s).(", old_cp->rval);
+                BufferAppend(grow, RvalScalarValue(rval), strlen(RvalScalarValue(rval)));
+                BufferAppendChar(grow, ')');
+                rval.item = xstrdup(BufferData(grow)); // TODO: will this leak?
```

Comment:
Yes, I think so. And looking at how ``Rval`` works I wouldn't even expect this to work correctly. ``Rval`` is passed by value, so if you change the ``rval`` variable here, the ``cp`` that you created further up should still have the old ``Rval``. On the other hand I might be confused, does this work for you?

As for the memory leak, what you can do is the following:

<pre>
...
BufferAppendChar(grow, ')');
RvalDestroy(rval);
rval.item = BufferClose(grow); // Avoids xstrdup.
</pre>